cache/server.go

```diff
@@ -38,24 +39,53 @@ func (s *RawProviderServer) ValidateDataSourceConfig(ctx context.Context, req *t
 	return resp, nil
 }
 
-// UpgradeResourceState isn't really useful in this provider, but we have to loop the state back through to keep Terraform happy.
 func (s *RawProviderServer) UpgradeResourceState(ctx context.Context, req *tfprotov5.UpgradeResourceStateRequest) (*tfprotov5.UpgradeResourceStateResponse, error) {
 	resp := &tfprotov5.UpgradeResourceStateResponse{}
 	resp.Diagnostics = []*tfprotov5.Diagnostic{}
 
-	sch := GetProviderResourceSchema()
-	rt := GetObjectTypeFromSchema(sch[req.TypeName])
+	var resourceValue tftypes.Value
+	var err error
 
-	rv, err := req.RawState.Unmarshal(rt)
-	if err != nil {
-		resp.Diagnostics = append(resp.Diagnostics, &tfprotov5.Diagnostic{
-			Severity: tfprotov5.DiagnosticSeverityError,
-			Summary:  "Failed to decode old state during upgrade",
-			Detail:   err.Error(),
-		})
-		return resp, nil
+	stateSchemaVersion := req.Version
```

Comment:
Need to check of the schema version of the prior state is different from the current schema version of the provider. If it's different, we need to update the prior state value to fit the new schema before we can proceed.